Improvements to StructureType support (towards #1031)

arporter opened the pull request:
This PR addresses a few issues found when processing NEMO with the latest master of PSyclone. In particular, the situations when we already have a generic Symbol for a derived type instead of a DataSymbol.

arporter commented:
This is ready for a first look. You'll see several TODO's that reference #935 but I'm not sure that's the correct issue any more so please feel free to educate me.

sergisiso commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -2585,7 +2616,8 @@ def _array_syntax_to_indexed(self, parent, loop_vars):
                     symbol = _find_or_create_imported_symbol(
-                                array, loop_vars[range_idx])
+                        array, loop_vars[range_idx],
+                        symbol_type=DataSymbol, datatype=DeferredType())
```
Can we infer that Range indexes are of Integer Datatype?

arporter replied:
Mmm, we could but we can't infer the precision so I think we should leave this. (Unless we can have an integer of unknown precision?)

sergisiso replied:
After I wrote the comment I thought that even if we know the datatype it expects it may be possible to have some implicit casts to that type, so symbol datatype inference may be very tricky. The inference of the symbol_type is much easier because it just depends on where the symbol it is found. So I agree with your decision to leave it as DeferedType.

sergisiso reviewed (changes requested):
@arporter I made some in-line comments about the code, the PR looks almost ready. Its good to see that all these "disable_declaration_checks" can be removed when adding generic/deferred symbol declarations.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -163,7 +163,14 @@ def _find_or_create_imported_symbol(location, name, scope_limit=None,
                 # return the Symbol.
-                return symbol_table.lookup(name, scope_limit=test_node)
+                sym = symbol_table.lookup(name, scope_limit=test_node)
+                if "symbol_type" in kargs and type(sym) == Symbol:
```
That makes sense; I've made the suggested change. I've also updated the docstring which I forgot to do last time.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -163,7 +163,14 @@ def _find_or_create_imported_symbol(location, name, scope_limit=None,
+                if "symbol_type" in kargs and type(sym) == Symbol:
+                    # The caller specified a sub-class of Symbol and we've
+                    # only found a generic Symbol.
+                    # TODO 935
```
Done, thanks.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -1078,13 +1085,16 @@ def _parse_dimensions(dimensions, symbol_table):
-                  dimension is unknown, otherwise it holds an integer \
-                  with the extent. If it is an empty list then the symbol \
-                  represents a scalar.
+            index is contiguous in memory). Each entry represents an array \
```
Oops.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -1116,8 +1126,14 @@ def _unsupported_type_error(dimensions):
                         sym = symbol_table.lookup(dim_name)
-                        if isinstance(sym.datatype, (UnknownType,
-                                                     DeferredType)):
+                        # pylint: disable=unidiomatic-typecheck
```
Updated and yes, that's right.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -1116,8 +1126,14 @@ def _unsupported_type_error(dimensions):
+                        # pylint: disable=unidiomatic-typecheck
+                        if type(sym) == Symbol:
+                            # TODO #935 convert this to a DataSymbol
+                            sym.__class__ = DataSymbol
```
Yes, good point. I meant to update the docstrings and then forgot.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -1437,7 +1453,7 @@ def _process_decln(self, parent, symbol_table, decl, default_visibility,
         elif isinstance(type_spec, Fortran2003.Declaration_Type_Spec):
             # This is a variable of derived type
-            type_name = str(walk(type_spec, Fortran2003.Type_Name)[0])
+            type_name = str(walk(type_spec, Fortran2003.Type_Name)[0]).lower()
```
It was necessary but actually, as discussed in Teams, this was because of a bug in `SymbolTable.remove()`. I've fixed the bug, added tests and removed this `lower()`.

arporter commented:
Ready for another look now.

sergisiso commented on src/psyclone/psyir/nodes/array_of_structures_reference.py:
```diff
@@ -96,11 +96,16 @@ def create(symbol, indices, members, parent=None):
-        if not isinstance(symbol.datatype, symbols.ArrayType):
+        if isinstance(symbol.datatype, symbols.ArrayType):
+            base_type = symbol.datatype.intrinsic
+        elif isinstance(symbol.datatype, (symbols.DeferredType,
```
Update the method docstring to say that DeferredType and UnknownType are accepted

arporter replied:
Done.

sergisiso reviewed (changes requested):
@arporter This is almost ready. Pylint and pycodestyle pass without new issues. The reference guide builds the new auto-reference without new issues.

The test shows 100% covered, but could you add some explicit test calling _find_or_create_imported_symbol and checking some symbol type changes. I think it is just checked indirectly at the moment if I haven't miss it.

Also can you provide and issue number this PR closes or goes towards.

arporter commented:
I've moved the existing tests for _find_or_create_imported_symbol() into a separate file and added one for the symbol-type changing functionality. This PR is sort-of a small step towards #1031 in that it relaxes the requirement that the type must be a `TypeSymbol`. It does not however yet permit it to be a `StructureType`.
CI-permitting, this is ready for another look.

sergisiso reviewed (approved):
@arporter All comments addressed. The tests for new functionality are now explicit and pass (also with compilation). This PR is now ready to be merged.
